### PR TITLE
Iteration 2 ross refactor

### PR DESCRIPTION
**Pull request by perryr16**

### CODE REVIEW

File: lib/game.rb [hunk 1]
```diff
@@ -14,14 +14,60 @@ def self.from_csv(csv_file_path)
...
+    end
+  end
+#one of THE METHODS ross
+  def self.average_goals_by_season
+    goals = home_away_goals_per(:season)
+    season_length = games_per(:season)
+    average_goals_per_season = average_goals(goals, season_length)
+    season_ids = @@all.map { |game| game.season}.uniq
+    Hash[season_ids.zip(average_goals_per_season)]
+  end
```
Comment (perryr16) [inline lib/game.rb:59]: This method is able to pull 4 deliverables 
highest scoring visitor
highest scoring home team
lowest scoring visitor
lowest scoring home team

File: lib/game.rb [hunk 2]
```diff
@@ -14,14 +14,60 @@ def self.from_csv(csv_file_path)
...
+
+  def self.home_away_goals_per(header)
+    goals_per(header, :away_goals) + goals_per(header, :home_goals)
+  end
+
+  def self.average_goals(sum_array, length_array)
+    sum_array.each_with_index.map do |goals, index|
+      (goals.to_f / length_array[index].to_f ).round(2)
+    end
+  end
```
Comment (perryr16) [inline lib/game.rb:51]: These commented methods are the deliverables. we can delete them but it helped me distinguish between helpers and deliverables

File: lib/stat_tracker.rb [hunk 3]
```diff
@@ -60,5 +60,35 @@ def percentage_ties
   def count_of_games_by_season
     Game.count_of_games_by_season
   end
```
Comment (perryr16) [inline lib/stat_tracker.rb:62]: again, I just added #ross for my sanity

File: test/fixtures/short_game_teams.csv [hunk 4]
```diff
@@ -19,6 +19,8 @@ game_id,team_id,HoA,result,settled_in,head_coach,goals,shots,tackles,pim,powerPl
 2012030314,6,home,WIN,REG,Claude Julien,1,6,25,8,3,0,52.5,11,5
 2012030231,17,away,LOSS,REG,Mike Babcock,1,5,43,10,3,0,52.5,6,4
 2012030231,16,home,WIN,REG,Joel Quenneville,2,10,24,10,3,1,47.5,2,11
```
Comment (perryr16) [inline test/fixtures/short_game_teams.csv:21]: we'll have to sort this out.

File: lib/game.rb (hunk 1 above)
Comment (stellakunzang) [reply lib/game.rb:59]: I'm sure this is all very smart but I don't understand any of it

**Comment by foleymichelle9** [comment]
These all look great! You're finding a lot of averages so i'm finding a good use case for a calcuable module!

File: test/game_test.rb [hunk 5]
```diff
@@ -61,6 +62,40 @@ def test_it_has_all
...
+
+  def test_games_per_season
+    assert_equal [2, 5, 6, 4, 2, 1], Game.games_per(:season)
+  end
+
+  def test_home_and_away_goals_per_season
+    assert_equal [4, 10, 13, 10, 4, 3], Game.goals_per(:season, :away_goals)
+    assert_equal [6, 12, 11, 8, 5, 2], Game.goals_per(:season, :home_goals)
+  end
+
```
Comment (perryr16) [inline test/game_test.rb:77]: lots of passed arguments

File: test/game_test.rb [hunk 6]
```diff
@@ -61,6 +62,40 @@ def test_it_has_all
...
+  def test_average_goals_per_header
+    goals = [10, 20]
+    season_lengths = [3, 8]
+    assert_equal [3.33, 2.5], Game.average_goals(goals, season_lengths)
+  end
+
+  def test_average_goals_by_season
+    Game.all.stubs(:map).returns([20122013, 20162017])
```
Comment (stellakunzang) [inline test/game_test.rb:93]: nice stubs

File: lib/game.rb [hunk 7]
```diff
@@ -14,14 +14,60 @@ def self.from_csv(csv_file_path)
...
+    sum = @@all.sum { |game| game.away_goals + game.home_goals}.to_f
+    (sum / @@all.length.to_f).round(2)
+  end
+
+  def self.games_per(header)
+    group_by_header = @@all.group_by { |game| game.send(header) }
+    group_by_header.values.map{ |games| games.length}
+  end
+
+
```
Comment (perryr16) [inline lib/game.rb:35]: I ended up passing a lot of arguments, trying to cut down on lines of code

File: test/game_test.rb [hunk 8]
```diff
@@ -61,6 +62,40 @@ def test_it_has_all
...
+    assert_equal [2.0, 2.0, 2.17, 2.5, 2.0, 3.0], Game.average_goals(away_goals, games)
+    assert_equal [3.0, 2.4, 1.83, 2.0, 2.5,2.0], Game.average_goals(home_goals, games)
+  end
+
+  def test_average_goals_per_header
+    goals = [10, 20]
+    season_lengths = [3, 8]
+    assert_equal [3.33, 2.5], Game.average_goals(goals, season_lengths)
+  end
+
```
Comment (perryr16) [inline test/game_test.rb:91]: STUBS!

File: test/stat_tracker_test.rb [hunk 9]
```diff
@@ -69,4 +70,85 @@ def test_it_can_find_percentage_ties
...
+    Game.all.stubs(:map).returns([1, 2, 3, 4])
+    #home_goals
+    Game.stubs(:goals_per).returns([100, 60, 10, 20])
+    #games
+    Game.stubs(:games_per).returns([10, 10, 10, 10])
+    #lowest_scoring_visitor_team_
+    assert_equal "Houston Dynamo", @stat_tracker.lowest_scoring_visitor
+  end
+
+
```
Comment (perryr16) [inline test/stat_tracker_test.rb]: should delete this commented code, was used as reference

**Review by perryr16 — COMMENTED** [review]
my piece

File: test/game_test.rb [hunk 10]
```diff
@@ -72,4 +107,29 @@ def test_it_can_calculate_lowest_total_score
...
     assert_equal ({20122013=>2, 20162017=>5, 20142015=>6, 20132014=>4, 20152016=>2, 20172018=>1}), Game.count_of_games_by_season
   end
+
+  def test_nth_scoring_visitor_or_home_by_id
+    #team_ids
+    Game.all.stubs(:map).returns([1, 2, 3, 4])
+    #away_goals
+    Game.stubs(:goals_per).returns([80, 60, 40, 20])
+    #games
+    Game.stubs(:games_per).returns([10, 10, 10, 10])
```
Comment (foleymichelle9) [inline test/game_test.rb:117]: Good use of stubs here!

**Comment by stellakunzang** [comment]
Excellent, of course. I vote in favor of merging!

File: test/fixtures/short_game_teams.csv [hunk 11]
```diff
@@ -19,6 +19,8 @@ game_id,team_id,HoA,result,settled_in,head_coach,goals,shots,tackles,pim,powerPl
...
+<<<<<<< HEAD
+=======
```
Comment (foleymichelle9) [inline test/fixtures/short_game_teams.csv:23]: Could probably delete these lines

**Comment by foleymichelle9** [comment]
Agreed! Looks great!

File: lib/game.rb (hunk 1 above)
Comment (stellakunzang) [reply lib/game.rb:59]: https://media.giphy.com/media/FmBhzktIjvdZe/giphy.gif

**Comment by Rostammahabadi** [comment]
Looks good!

**Pull request merged**